design/reboot-interface.md

```diff
@@ -0,0 +1,155 @@
+
+A declarative API is proposed to request the baremetal operator to reboot a
+provisioned Host in an atomic step (i.e. without the need to make multiple
+sequential changes to the spec).
+
+## Motivation
+
+We require a non-destructive way to fence a Kubernetes node. Some nodes cannot
+be replaced (it is currently not possible for a new master to join a cluster),
+or are expensive to replace (e.g. if this would require rebalancing Ceph data).
+A solution to this is for fencing to reboot the Host, thus ensuring that
+running processes are stopped to avoid a split-brain scenario while still
+allowing the node to rejoin the cluster with its data intact (albeit stale)
+after the reboot.
+
+The expected implementation of this on the fencing side is a Custom Resource
+Definition that requests a Machine be rebooted. The actual reboot will be
+effected either by the Machine Actuator itself (if the Cluster-API SIG can be
+persuaded that this should be part of the API in the long term), or some
+equivalent of the Machine Actuator limited to just this purpose.
+
+### Goals
+
+The Machine (Reboot) Actuator will require from the BareMetalHost:
+
+* A declarative API to perform the reboot
+* The ability to determine a point where all processes running on the Host at
+  the time of the fencing decision are guaranteed to have been stopped
+* Rapid convergence to the fenced state independent of other happenings in the
+  system
+
+### Non-Goals
+
+There is no requirement to implement a scheduled reboot. In the Kubernetes
+context, reboot decisions should generally be made in realtime by some higher
+level of the Node/Machine hierarchy, to take into account such questions as the
+overall health of the cluster and the effect of a reboot on that. In any event,
+the best implementation for this would be a separate RebootSchedule CRD that
+waits until the appointed time before issuing an immediate reboot request to
+the BareMetalHost. This allows multiple reboots to be scheduled, scheduled
+reboots to be manipulated or cancelled, and a record to be left behind of past
+scheduled reboots. Any such resource would have the same requirements of the
+BareMetalHost as fencing does.
+
+This API is not responsible for managing unprovisioned hosts, e.g. to recover
+from state inconsistencies in Ironic prior to provisioning. Any such
+inconsistencies that are not handled internally by the BareMetalHost will
+likely need manual intervention to recover anyway.
+
+## Proposal
+
+A new date-time field, ``lastBooted``, will be added to the ``provisioning``
+section of the BareMetalHost status. This records a time after which the Host
+was last booted using the current image. Processes running prior to this time
+may be assumed to have been fenced.
+
+To provide a declarative interface for performing a reboot, a new
+``rebootRequestTimestamp`` field will to be added to the BareMetalHost spec.
+This field could have any type, so long as clients are able to provide a unique
```

Comment:
If I provide unique strings as my values, that are not timestamps, how does it know when I've modified it to a different value?

Reply:
What this is intending to say is that we could have designed it to take any unique string, but instead we designed it to take a timestamp (which is a unique string with some other nice properties).
Validation will automatically enforce that the user provides something resembling a timestamp here.